src/toil/batchSystems/kubernetes.py

```diff
@@ -496,47 +492,126 @@ def _getUpdatedBatchJobImmediately(self):
+                    pod = self._getPodForJob(j)
+
+                    if pod is None:
+                        # Skip jobs with no pod
+                        continue
+
+                    waitingInfo = getattr(pod.status.container_statuses[0].state, 'waiting')
+                    if waitingInfo is not None and waitingInfo.reason == 'ImagePullBackOff':
+                        # Assume it will never finish, even if the registry comes back or whatever.
+                        # We can get into this state when we send in a non-existent image.
+                        # See https://github.com/kubernetes/kubernetes/issues/58384
+                        jobObject = j
+                        chosenFor = 'stuck'
+                        logger.warning('Failing stuck job; did you try to run a non-existent Docker image?'
+                                       ' Check TOIL_APPLIANCE_SELF.')
+                        break
+           
+            w = kubernetes.watch.Watch()    
+            if jobObject is None:
+                # Nothing is ready yet.
+
+                if self.enableWatching:
+                    for j in self._ourJobObjects():
+                        print(j.spec.template.metadata.labels[u'job-name'], type(j.spec.template.metadata.labels[u'job-name']))
+                        for event in w.stream(self.coreApi.list_namespaced_pod, self.namespace, timeout_seconds=maxWait):
+                            job = event['object']
+                            if job.metadata.name.startswith(self.jobPrefix):
+                                logger.info("Event: %s %s %s" % (event['type'],event['object'].kind, event['object'].metadata.name))
+                                if job.status.phase == 'Failure':
+                                    logger.info("FAILED")
+                                    logger.warning(job.status.container_status[0].state.reason,
+                                        job.status.container_statuses[0].state.terminated.exit_code)
+                                    return None
+                                elif job.status.phase == 'Succeeded':
+                                    logger.info("SUCCEEDED NAME: %s EXIT CODE: %s TOTAL TIME: %s" % (job.metadata.name,\
+                                        job.status.container_statuses[0].state.terminated.exit_code, \
+                                        (job.status.container_statuses[0].state.terminated.finished_at - \
+                                        job.status.container_statuses[0].state.terminated.started_at).total_seconds()))
+                                    return (job.metadata.name,\
+                                        job.status.container_statuses[0].state.terminated.exit_code, \
+                                        (job.status.container_statuses[0].state.terminated.finished_at - \
+                                                job.status.container_statuses[0].state.terminated.started_at).total_seconds())
```

Comment:
Here you're returning the job's name as the first item, where you need to be returning the job's ID number (an int).

```
jobID = int(job.metadata.name[len(self.jobPrefix):])
```

You also have this big long tuple of things that reach deep into `job` repeated twice. You probably want to condense this down:

```
jobID = int(job.metadata.name[len(self.jobPrefix):])
terminated = job.status.container_statuses[0].state.terminated
runtime = (termianted.finished_at - terminated.started_at).total_seconds()
result = (jobID, terminated.exit_code, runtime)
```

And then use `result` instead of the giant expression in both places.

You also need to make sure to [delete the job that has finished, and wait for it to no longer exist](https://github.com/DataBiosphere/toil/pull/2848/files#diff-183d11f3af830162fadf362f8c641fdfR631-R650), before returning. Otherwise, the next time someone asks for an updated job, that job we already reported (when we saw it finish) will still be there in Kubernetes in the finished state, so we'll grab it and report it as having finished again, which we aren't supposed to do.